app/api/auth/[...nextauth]/route.ts

```diff
@@ -0,0 +1,107 @@
+import NextAuth, { type DefaultUser, type NextAuthOptions } from 'next-auth';
+import GithubProvider from 'next-auth/providers/github';
+import GoogleProvider from 'next-auth/providers/google';
+import CredentialsProvider from 'next-auth/providers/credentials';
+
+import User, { IUser } from '@/models/User';
+import { connectToMongoDB } from '@/lib/database/mongodb';
+
+// ----------------------------------------------------------------
+
+const authOptions: NextAuthOptions = {
+  providers: [
+    GithubProvider({
+      clientId: process.env.GITHUB_ID ?? '',
+      clientSecret: process.env.GITHUB_SECRET ?? '',
+    }),
+    GoogleProvider({
+      clientId: process.env.GOOGLE_ID ?? '',
+      clientSecret: process.env.GOOGLE_SECRET ?? '',
+    }),
+    CredentialsProvider({
+      name: 'signin',
+      id: 'signin',
+      credentials: {
+        email: {
+          label: 'Email',
+          type: 'text',
+          placeholder: 'Enter your email address',
+        },
+        password: { label: 'Password', type: 'password' },
+      },
+      async authorize(credentials, req) {
+        console.log('credentials', credentials);
+
+        let user: IUser | null = null;
+
+        try {
+          await connectToMongoDB();
+
+          // ? Should the password be encrypted when
+          user = await User.findOne({
+            email: credentials?.email,
+            password: credentials?.password,
+          });
+
+          console.log('user from mongo', user);
+        } catch (error) {
+          console.log('Error fetching user from MongoDB', error);
+        }
+
+        // Have to check this as well since TS file for MongoDB says Document._id is optional ---> string | undefined
+        if (user) {
+          return {
+            ...user,
+            id: user?._id,
+          };
+        }
+
+        return null;
+      },
+    }),
+    CredentialsProvider({
+      name: 'signup',
+      id: 'signup',
+      credentials: {
+        fullname: {},
+        email: {
+          // label: 'Email',
+          // type: 'text',
+          // placeholder: 'Enter your email address',
+        },
+        password: {
+          //  label: 'Password', type: 'password'
+        },
+      },
+      async authorize(credentials, req) {
+        // console.log('credentials', credentials);
+        // const user: User = {
+        //   id: '1',
+        //   name: 'J Smith',
+        //   email: 'jsmith@example.com',
+        //   image: '',
+        // };
+        return null;
+        // return user;
+        // return null;
+      },
+    }),
+  ],
```

Comment:
The Credentials Provider is meant to receive credentials, and return the user that would indicate to NextAuth that you're logged in. While in theory this would work with a separate Sign Up Credentials Provider ( which would register and log the user in ), it would also make sense to just have a normal API endpoint or server action, that creates the user account in the database and redirects the user to the login page.